### DISCUSSION TITLE
Production error

### DISCUSSION THREAD

**Discussion by diego-rapoport** [TypedProject/ts-express-decorators#1953] [Q&A] [answered]
Hi again! I'm trying to run my application in production mode. The build went fine after I corrected all the errors and it seemed that everything would work.
But then I got this error:
`{"startTime":"2022-06-22T15:08:07.950Z","categoryName":"TSED","level":"ERROR","errno":-2,"syscall":"open","code":"ENOENT","data":[]}`

No further explanations. Where can I see a better log for this error?
I ran `npm run start:prod` after the build

**Comment by Romakita** [comment, marked as answer]
Hello, 
The stack doesn’t help. Are on the latest version. Also if it’s an issue, it’s preferable to open issue ;)

your error reflect a missing modules when the componentsScan import a module. 

The componentsScan is a real problem. It preferable to use direct import or use barrelspy to generate import from globpattern (the latest cli generate a project with barrelsby and componentsScan disabled). 

For your actuall problem tell me if you are one the latest version.

**Reply by Romakita** [reply]
Out-of-the-box, the `npm start:prod` command works. So your issue is relation to your code implementation.
This why it's really important to test the production mode in your continuous delivery chain (CI/CD).

There I have the impression that you have massively developed the application and that you realize that there is a problem with the production profile. And now you are asking me why, when this is not my app. It is in these moments that I should offer you a commercial offer to provide you with support ;) 

There a simple solution to have the correct log in this case ;)

**Reply by diego-rapoport** [reply]
My version is 6.114.17. I don't really know how to use barrelspy but I guess that imports have been an issue around my project. Even maybe a wrong configuration of `tsconfig.json` with my pattern's import.

I could really appreciate a direct support from you, but be aware that we are a small company from Brazil and I'm basically the only programmer around. So yes, I could take an offer and see if it fits for us because we need to solve the problem as soon as possible.

**Reply by Romakita** [reply]
I understand @diego-rapoport ;)
Sorry for the somewhat harsh message. I receive a lot of support requests which can usually be solved without my help.

The problem come from the tsed logger that not display the error correctly on the bootstrap server (I haven't fixed that because I'm not able to reproduce correctly this problem). But a simple solution is to log the error with a class console.error():

```typescript
import { $log } from "@tsed/common";
import { PlatformExpress } from "@tsed/platform-express";
import { Server } from "./Server";

async function bootstrap() {
  try {
    $log.debug("Start server...");
    const server = await PlatformExpress.bootstrap(Server);

    await server.listen();
    $log.debug("Server initialized");
  } catch (er) {
    console.log(er); /// HERE !
  }
}

bootstrap();

```
If this solution doesn't display the error correctly tell me and we'll investigate your problem together (via slack if possible).

See you
Romain

**Reply by diego-rapoport** [reply]
Thank you very much for the support and reply!! So, in my code I already have a `$log.error(er)` where I should put a `console.log(er)`. Does that make total difference?

**Reply by Romakita** [reply]
$log.error seems not to display the error correctly in the terminal. I hope the console.log will works better because it doesn’t try to transform the error object

**Reply by Romakita** [reply]
So it change something?

**Reply by diego-rapoport** [reply]
I'm reviewing something that could also made that error. Perhaps by the end of the noon I could give you a proper answer.

**Reply by diego-rapoport** [reply]
I was having a mongodb connection problem earlier, but now I have a new error and it is better explicit:
```
ClienteService->constructor(modelo: Object)
                       ^‾‾‾‾‾
    at Object.getMetadata (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/reflect-metadata/Reflect.js:354:23)
    at Function.getParamTypes (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/core/lib/cjs/domain/Metadata.js:468:24)
    at Map.mapInvokeOptions (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:583:44)
    at Map.resolve (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:499:35)
    at Map.invoke (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:171:29)
    at /home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:514:101
    at Array.map (<anonymous>)
    at Map.resolve (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:519:35)
    at Map.invoke (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:180:103)
    at Map.loadSync (/home/dinhego/Programação/Angular/Projetos/painelxdigital/serverTs/node_modules/@tsed/di/lib/cjs/services/InjectorService.js:219:22) {
  name: 'INJECTION_ERROR',
  tokens: [ [class ClienteService extends GenericService] ],
  origin: TypeError: Unable to inject dependency.
```

**Reply by diego-rapoport** [reply]
Hi @Romakita, sorry if I'm disturbing you. If there's any idea about how to solve this problem, or where I could find something to solve this, it would be greatly appreciate. Otherwise I should find another way to implement this.

Thank you in advance for all the support already given by you.

**Reply by Romakita** [reply]
Sorry i thought you found the problem :)

In this case, isn't easy to know exactly what is the issue (probably a circular ref - but you should have the same issue in dev mode). 

The following message indicate that you have an injected class with type (because it's resolved as an Object and not to as class).

```
ClienteService->constructor(modelo: Object)
```

This problem occurs when you have a circular references between two or more service. Detect and solve this problem is hard. You can confirm this diagnostic by addind a console before the class ClienteService:
```typescript
import {ModeloService} from "../services/Modelo";

console.log(ModeloService); // if `{}` it's a circular reference (don't forget to build your code)

class ClienteService {
  constructor(modelo: ModeloService) {}
}
```

A quick solution is to use `@Inject(() => ModelService)`:

```typescript
class ClienteService {
  @Inject(() => ModelService)
  modelo: ModeloService

  constructor()
}
```

But it's preferable to found the root cause of your circular ref and fix it before using `@Inject(() => ModelService}`

If you are not able to solve your issue, I can take a look, but I need a repo access if it's possible. 

You'll also find my sponsorship page in case you want me to investigate on your project and find a solution: https://github.com/sponsors/Romakita?frequency=one-time&sponsor=Romakita

See you
Romain

**Reply by diego-rapoport** [reply]
Just for the record, I'm not having this issue in dev mode and that's why I found it so odd. 
What I'm doing there is that I have a generic class as a service so I can extend it from other services using a specific model o the constructor. 

So my class from `ClienteService` is basically:
```typescript
@Injectable()
export class ClienteService extends GenericService {
  constructor(modelo: MongooseModel<Clientes>) {
    super(modelo)
  }
}
```
I don't know if that problem will also appear on all the other services extending from `GenericService`, but I'd assume so. Maybe my problem is how I implemented the `GenericService`:
```typescript
export class GenericService {

  id: number;

  constructor(
    private modelo: MongooseModel<any>,
    private contador?: MongooseModel<ContadoresModel>
  ) { }

  // TODO: Transformar o findById numa func separada
  async query(opts: any) {
    if (opts.param && isValidObjectId(opts.param)) {
      if (isValidObjectId(opts.param)) {
        const resultado = await this.modelo.findById(opts.param)
        return resultado
      }

      const resultado = await this.modelo.findOne(opts.param)
      return resultado
    }
    const resultado = await this.modelo.findOne(opts)
    return resultado
  }

  async queryMany(opts: any) {
    if (!opts) {
      const resultado = await this.modelo.find()
      return resultado
    }
    if (opts.param && isValidObjectId(opts.param)) {
      const resultado = await this.modelo.findById(opts._id)
      return resultado
    } else if (opts) {
      const resultado = await this.modelo.find(opts)
      return resultado
    }
  }

  async salva(params: any) {
    if (this.contador) {
      const atualizado = await this.contador.updateOne({ tipo: params.tipo }, { $inc: { contador: 1 } }, { new: true })
      // TODO: Log do atualizado
      const contador = await this.contador.findOne({ tipo: params.tipo })
      if (contador) this.id = contador.contador
    }

    const novoDoc = new this.modelo({ ...params, id: this.id })
    const resultado = await novoDoc.save()

    // return novoDoc;
    return resultado
  }

  async atualiza(params: { query: { [key: string]: string }, atualizar: { [key: string]: string | number } }) {
    // TODO: generalizar func
    const resultado = await this.modelo.findOne(params.query).then(async (res: any) => {
      if (res !== null && res !== undefined) {
        res = Object.assign(res, params.atualizar)
        await res!.save()
        return res!.toClass()
      }

    })
      .catch((e: any) => console.log('ERRO NO FINDONE = ', e))

    return resultado
  }
}
```

Does that make any sense?

**Reply by Romakita** [reply]
It should not works. You are just lucky... (or not).

this code in typescript won't generates the correct Metadata type, because it's a generics:

```
modelo: MongooseModel<Clientes> ==> Metadata Object
```

THIS IS WHY you'll never find IN THE OFFICIAL Ts.ED DOCUMENTATION ^^ an example to inject mongoose model without `@Inject()` annotation.

But solving your issue is really simple!

```typescript
export class GenericService<Model, Cantador = ContadoresModel> {
  id: number;
  protected modelo: MongooseModel<Model>,  // give generic here will improve your code completion and type checking!
  protected contador?: MongooseModel<Cantador>;
}
```

Then:

```typescript
@Injectable()
export class ClienteService extends GenericService<Clientes> {
  @Inject(Clientes)
  protected modelo: MongooseModel<Clientes>;
  
  // @Inject(CantardoresModel)
  // protected contador?: MongooseModel<CantardoresModel>;
}
```

For me,  it's really strange that your code works as expected in dev. Typescript generates exactly the same metadata in dev and in prod. Ts.ED doesn't do magic on this part (not yet ^^).

See you
Romain

**Reply by diego-rapoport** [reply]
@Romakita thank you soooooo much for your support. It really helped me and made it all work. Now I just have to finish the details, but the building is working.
Will sponsor you soon, for sure! Great framework, by the way.

Cheers from Brazil.
